lib/src/repo.rs

```diff
@@ -209,21 +213,28 @@ impl ReadonlyRepo {
     pub fn readonly_index(&self) -> &dyn ReadonlyIndex {
         self.index
             .get_or_init(|| {
-                self.index_store
-                    .get_index_at_op(&self.operation, &self.store)
+                Box::into_pin(
+                    self.index_store
+                        .get_index_at_op(&self.operation, &self.store),
+                )
             })
-            .as_ref()
+            .deref()
     }
 
-    fn change_id_index(&self) -> &ChangeIdIndex {
-        self.change_id_index.get_or_init(|| {
-            let heads = self.view().heads().iter().cloned().collect_vec();
-            let walk = self.readonly_index().as_index().walk_revs(&heads, &[]);
-            IdIndex::from_vec(
-                walk.map(|entry| (entry.change_id(), entry.position()))
-                    .collect(),
-            )
-        })
+    fn change_id_index<'a>(&'a self) -> &'a (dyn ChangeIdIndex + 'a) {
+        let change_id_index: &'a (dyn ChangeIdIndex + 'a) = self
+            .change_id_index
+            .get_or_init(|| {
+                let revset: Box<dyn Revset<'a>> = RevsetExpression::all().evaluate(self).unwrap();
+                let change_id_index: Box<dyn ChangeIdIndex + 'a> = revset.change_id_index();
+                let change_id_index: Box<dyn ChangeIdIndex> =
+                    unsafe { std::mem::transmute(change_id_index) };
+                change_id_index
+            })
+            .as_ref();
+        let change_id_index: &'a (dyn ChangeIdIndex + 'a) =
+            unsafe { std::mem::transmute(change_id_index) };
```

Comment:
Unneeded unsafe cast

Reply:
Good catch. (I think this was needed before I removed the lifetime parameter to `ChangeIdIndex`.)